srcs/app/CGI/CgiHandler.cpp

```diff
@@ -26,21 +37,217 @@ ssize_t	CgiHandler::SendCgiOutput() {
 	if (nbytes > 0) {
 		data_.erase(0, nbytes);
 	}
+	if (data_.empty()) {
+		fd_sets_->removeFd(socket_);
+	}
 	return nbytes;
 }
 
-bool	CgiHandler::HasDataAvailable() const {
-	return !data_.empty();
-}
-
 int		CgiHandler::GetSocket() const {
 	return socket_;
 }
 
 int		CgiHandler::GetCgiOutFd() const {
-	return cgi_output_;
+	return cgi_info_.cgi_output_fd;
+}
+
+bool	CgiHandler::IsComplete() const {
+	return cgi_complete_ && data_.empty();
+}
+
+void	CgiHandler::TryParseHeaders_() {
+	try {
+		ParseHeaders_();
+		if (cgi_complete_ && !headers_parsing_complete_) {
+			throw std::runtime_error("Invalid header");
+		}
+	}
+	catch (const std::exception &) {
+		fd_sets_->removeFd(cgi_info_.cgi_output_fd);
+		SetErrorResponse_(500);
+		if (!cgi_complete_) {
+			kill(cgi_info_.pid, SIGKILL);
+		}
+		cgi_complete_ = true;
+	}
+}
+
+void	CgiHandler::ParseHeaders_() {
```

Comment:
As you know, there is a lot of common code between this one and srcs/app/http/HttpRequest.cpp:243 void	HttpRequest::ParseHeaders_(const std::string &raw_request);
Did you see it to be too complicated, timeconsuming or generally bothersome to factor it out?

Reply:
Thanks for pointing this out.

We could refactor the duplicate functionality of parsing HTTP headers to a new class/component. I decided not to do it in this PR because I thought it was time-consuming.

 I will add a new issue to track this.

Reply:
I have an opinion about this, that I would like to share.

We're programming in cpp, and we're trying to do the most object oriented code we can. Object orientation is programming paradigm that bundles code to data, and very often it is very much convenient, makes code easily extensible, clear ownership, and so on.

But I don't think it's a magic wand that solves every programming problem. In this instance, I think an acceptable solution would be to separate code from data, make the methods in the class that parses http headers static, and pass whatever data it needs on the arguments of the function itself, C style. Like that we could easily use it from inside either of the classes that need it, and I think the work footprint would be much smaller.

What do you think about this solution?

Reply:
Sounds good to me. I'll give it a try.

Reply:
If you want I can do it myself, but not on the schedule we were thinking about (before next Sunday). I could probably do most of it on Saturday, as I travel.

Reply:
I'll do it because I want to improve at refactoring and OOP. Thank you!